Assets/HoloToolkit/Input/Scripts/InputManager.cs

```diff
@@ -0,0 +1,581 @@
+            if (modalInputStack.Count > 0)
+            {
+                var modalInput = modalInputStack.Peek();
+
+                // If there is a focused object in the hierarchy of the modal handler, start the event
+                // bubble there
+                GameObject focusedObject = GazeManager.Instance.HitObject;
+                if (focusedObject != null && focusedObject.transform.IsChildOf(modalInput.transform))
+                {
+
+                    if (ExecuteEvents.ExecuteHierarchy(focusedObject, eventData, eventHandler))
+                    {
+                        return;
+                    }
+                }
+                // Otherwise, just invoke the event on the modal handler itself
+                else
+                {
+                    if (ExecuteEvents.ExecuteHierarchy(modalInput, eventData, eventHandler))
+                    {
+                        return;
+                    }
+                }
+            }
+
+            // If event was not handled by modal, pass it on to the current focused object
+            if (GazeManager.Instance.HitObject != null)
+            {
+                bool eventHandled = ExecuteEvents.ExecuteHierarchy(GazeManager.Instance.HitObject, eventData, eventHandler);
+                if (eventHandled)
+                {
+                    return;
+                }
+            }
+
+            // If event was not handled by the focused object, pass it on to any fallback handlers
+            if (fallbackInputStack.Count > 0)
+            {
+                var fallbackInput = fallbackInputStack.Peek();
+                ExecuteEvents.ExecuteHierarchy(fallbackInput, eventData, eventHandler);
+            }
+        }
+
+        /// <summary>
+        /// Register to gaze manager events.
+        /// </summary>
+        private void RegisterGazeManager()
+        {
+            if (!isRegisteredToGazeChanges && GazeManager.Instance != null)
+            {
+                GazeManager.Instance.FocusedObjectChanged += GazeManager_FocusedChanged;
+                isRegisteredToGazeChanges = true;
+            }
+        }
+
+        /// <summary>
+        /// Unregister from gaze manager events.
+        /// </summary>
+        private void UnregisterGazeManager()
+        {
+            if (isRegisteredToGazeChanges && GazeManager.Instance != null)
+            {
+                GazeManager.Instance.FocusedObjectChanged -= GazeManager_FocusedChanged;
+                isRegisteredToGazeChanges = false;
+            }
+        }
+
+        private void GazeManager_FocusedChanged(GameObject previousObject, GameObject newObject)
+        {
+            if (disabledRefCount > 0)
+            {
+                return;
+            }
+
+            if (previousObject != null)
+            {
+                ExecuteEvents.ExecuteHierarchy<IFocusable>(previousObject, null, (inputHandler, eventData) => { inputHandler.OnFocusExit(); });
```

Comment:
Agreed that this might save a few allocations. I'd suggest making that optimization at a later time, should it become necessary. It wouldn't be a breaking change, since it's an internal implementation detail of InputManager, and I'd like to get this first version in.

Let me know if you disagree.